File: app/src/main/java/com/kimjisub/launchpad/midi/driver/DriverRef.java
package com.kimjisub.launchpad.midi.driver;

import com.kimjisub.manager.Log;

public abstract class DriverRef {

	// ============================================================================================= OnCycleListener

	private OnCycleListener onCycleListener = null;

	public DriverRef setOnCycleListener(OnCycleListener listener) {
		onCycleListener = listener;
		return this;
	}

	public interface OnCycleListener {
		void onConnected();

		void onDisconnected();
	}

	////

	public void onConnected() {
		if (onCycleListener != null)
			onCycleListener.onConnected();
	}

	public void onDisconnected() {
		if (onCycleListener != null)
			onCycleListener.onDisconnected();
	}


	// ============================================================================================= OnGetSignalListener

	private OnGetSignalListener onGetSignalListener = null;

	public DriverRef setOnGetSignalListener(OnGetSignalListener listener) {
		onGetSignalListener = listener;
		return this;
	}

	public interface OnGetSignalListener {
		void onPadTouch(int x, int y, boolean upDown, int velo);

		void onFunctionkeyTouch(int f, boolean upDown);

		void onChainTouch(int c, boolean upDown);

		void onUnknownEvent(int cmd, int sig, int note, int velo);
	}

	////

	public abstract void getSignal(int cmd, int sig, int note, int velo);

	void onPadTouch(int x, int y, boolean upDown, int velo) {
		Log.midiDetail("onPadTouch(" + x + ", " + y + ", " + upDown + ", " + velo + ")");
		if (onGetSignalListener != null)
			onGetSignalListener.onPadTouch(x, y, upDown, velo);
	}

	void onFunctionkeyTouch(int f, boolean upDown) {
		Log.midiDetail("onFunctionkeyTouch(" + f + ", " + upDown + ")");
		if (onGetSignalListener != null)
			onGetSignalListener.onFunctionkeyTouch(f, upDown);
	}

	void onChainTouch(int c, boolean upDown) {
		Log.midiDetail("onChainTouch(" + c + ", " + upDown + ")");
		if (onGetSignalListener != null)
			onGetSignalListener.onChainTouch(c, upDown);
	}

	void onUnknownEvent(int cmd, int sig, int note, int velo) {
		Log.midiDetail("onUnknownEvent(" + cmd + ", " + sig + ", " + note + ", " + velo + ")");
		if (onGetSignalListener != null)
			onGetSignalListener.onUnknownEvent(cmd, sig, note, velo);
	}

	// ============================================================================================= OnSendSignalListener

	private OnSendSignalListener onSendSignalListener = null;

	public DriverRef setOnSendSignalListener(OnSendSignalListener listener) {
		onSendSignalListener = listener;
		return this;
	}

	public interface OnSendSignalListener {
		void onSend(final byte cmd, final byte sig, final byte note, final byte velo);
	}

	////

	void onSend(final byte cmd, final byte sig, final byte note, final byte velo) {
		if (onSendSignalListener != null)
			onSendSignalListener.onSend(cmd, sig, note, velo);
	}

	void sendSignal(int cmd, int sig, int note, int velo) {
		onSend((byte) cmd, (byte) sig, (byte) note, (byte) velo);
	}

	void send09Signal(final int note, final int velo) {
		sendSignal((byte) 9, (byte) -112, (byte) note, (byte) velo);
	}

	void send11Signal(final int note, final int velo) {
		sendSignal((byte) 11, (byte) -80, (byte) note, (byte) velo);
	}

	public abstract void sendPadLED(int x, int y, int velo);

	public abstract void sendChainLED(int c, int velo);

	public abstract void sendFunctionkeyLED(int f, int velo);

	public abstract void sendClearLED();
}

File: app/src/main/java/com/kimjisub/launchpad/midi/driver/DriverRef.kt
package com.kimjisub.launchpad.midi.driver

import com.kimjisub.launchpad.tool.Log

abstract class DriverRef {

	// OnCycleListener

	private var onCycleListener: OnCycleListener? = null
	fun setOnCycleListener(listener: OnCycleListener?): DriverRef {
		onCycleListener = listener
		return this
	}

	interface OnCycleListener {
		fun onConnected()
		fun onDisconnected()
	}

	////

	fun onConnected() = onCycleListener?.onConnected()

	fun onDisconnected() = onCycleListener?.onDisconnected()

	// OnReceiveSignalListener

	private var onReceiveSignalListener: OnReceiveSignalListener? = null
	fun setOnGetSignalListener(listener: OnReceiveSignalListener?): DriverRef {
		onReceiveSignalListener = listener
		return this
	}

	interface OnReceiveSignalListener {
		fun onReceived(cmd: Int, sig: Int, note: Int, velocity: Int)
		fun onUnknownReceived(cmd: Int, sig: Int, note: Int, velocity: Int)

		fun onPadTouch(x: Int, y: Int, upDown: Boolean, velocity: Int)
		fun onChainTouch(c: Int, upDown: Boolean)
		fun onFunctionKeyTouch(f: Int, upDown: Boolean)
	}

	////


	open fun getSignal(cmd: Int, sig: Int, note: Int, velocity: Int) {
		onReceiveSignalListener?.onReceived(cmd, sig, note, velocity)
	}

	fun onPadTouch(x: Int, y: Int, upDown: Boolean, velocity: Int) {
		onReceiveSignalListener?.onPadTouch(x, y, upDown, velocity)
	}

	fun onChainTouch(c: Int, upDown: Boolean) {
		onReceiveSignalListener?.onChainTouch(c, upDown)
	}

	fun onFunctionKeyTouch(f: Int, upDown: Boolean) {
		onReceiveSignalListener?.onFunctionKeyTouch(f, upDown)
	}

	fun onUnknownReceived(cmd: Int, sig: Int, note: Int, velocity: Int) {
		onReceiveSignalListener?.onUnknownReceived(cmd, sig, note, velocity)
	}

	// OnSendSignalListener

	private var onSendSignalListener: OnSendSignalListener? = null
	fun setOnSendSignalListener(listener: OnSendSignalListener?): DriverRef {
		onSendSignalListener = listener
		return this
	}

	interface OnSendSignalListener {
		fun onSend(cmd: Byte, sig: Byte, note: Byte, velocity: Byte)
		fun onSendRaw(messages: List<ByteArray>, cableNumber: Int)
	}

	////

	internal fun sendSignal(cmd: Byte, sig: Byte, note: Byte, velocity: Byte) {
		onSendSignalListener?.onSend(cmd, sig, note, velocity)
	}

	internal fun sendSignal(cmd: Int, sig: Int, note: Int, velocity: Int) {
		sendSignal(cmd.toByte(), sig.toByte(), note.toByte(), velocity.toByte())
	}

	internal fun sendRawSignal(bytes: ByteArray, cableNumber: Int = 0) {
		onSendSignalListener?.onSendRaw(listOf(bytes), cableNumber)
	}

	internal fun sendRawSignals(messages: List<ByteArray>, cableNumber: Int = 0) {
		onSendSignalListener?.onSendRaw(messages, cableNumber)
	}

	open fun initialize() {}

	/** Returns SysEx messages and cable number for initialization, or null if none needed. */
	open fun getInitSysEx(): Pair<List<ByteArray>, Int>? = null

	abstract fun sendPadLed(x: Int, y: Int, velocity: Int)
	abstract fun sendChainLed(c: Int, velocity: Int)
	abstract fun sendFunctionKeyLed(f: Int, velocity: Int)
	abstract fun sendClearLed()
}